core/domain/exp_domain.py

```diff
@@ -2499,3 +2499,470 @@ def add_contribution_by_user(self, contributor_id):
+                    self.changed_properties[state_name].append(
+                        STATE_PROPERTY_WRITTEN_TRANSLATIONS)
+            else:
+                self.changed_properties[state_name] = [
+                    STATE_PROPERTY_WRITTEN_TRANSLATIONS]
+
+    def is_change_list_mergeable(
+            self, change_list,
+            exp_at_change_list_version, current_exploration):
+        """Checks whether the change list from the old version of an
+        exploration can be merged on the latest version of an exploration.
+
+        Args:
+            change_list: list(ExplorationChange). List of the changes made
+                by the user on the frontend, which needs to be checked
+                for mergeability.
+            exp_at_change_list_version: obj. Old version of an exploration.
+            current_exploration: obj. Exploration on which the change list
+                is to be applied.
+
+        Returns:
+            dict. A response dict for the change_list.
+            The response dict has two keys:
+                - 'changes_are_mergeable': boolean. Whether the given change
+                    list is mergeable on the current_exploration or not.
+                - 'send_email': boolean. Whether we need to send the change
+                    list to the admin to review for the future improvement
+                    of the cases to merge the change list.
+        """
+
+        # Old_to_new_state_names: dict. Dictionary mapping state names of
+        # prev_exp_version to the state names of current_exp_version.
+        # It doesn't include the name changes of added/deleted states.
+        old_to_new_state_names = {
+            value: key for key, value in self.new_to_old_state_names.items()
+        }
+
+        if (
+                len(self.added_state_names) > 0 or
+                len(self.deleted_state_names) > 0):
+            # In case of the addition and the deletion of the state,
+            # we are rejecting the mergebility because these cases
+            # change the flow of the exploration and are quite complex
+            # for now to handle. So in such cases, we are sending the
+            # changelist, frontend_version, backend_version and
+            # exploration id to the admin, so that we can look into the
+            # situations and can figure out the way if it’s possible to
+            # handle these cases.
+
+            return False, True
+
+        changes_are_mergeable = False
+
+        # state_names_of_renamed_states: dict. Stores the changes in
+        # states names in change_list where the key is the state name in
+        # frontend version and the value is the renamed name from the
+        # change list if there is any rename state change.
+        state_names_of_renamed_states = {}
+        for change in change_list:
+            change_is_mergeable = False
+            if change.cmd == CMD_RENAME_STATE:
```

Comment:
I tried doing this change in this PR only but the problem is that we'll need some extra variables like exp_at_change_list_version and current_exploration for every change and if we'll pass those variables then it will be just the same as is_change_list_mergeable() function, so I didn't change that. Will create an issue.

Reply:
#13373